{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create SRA submission sheet\n",
    "This Python Jupyter notebook creates a BioSample submission sheet for the Sequence Read Archive:\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import glob\n",
    "import itertools\n",
    "import os\n",
    "\n",
    "import natsort\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display full columns of data frames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the configuration for the uploads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('upload_config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the Illumina runs to submit:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Illumina runs from ./barcode_runs_upload.csv\n",
      "Here are the first few entries in the Illumina runs:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sort_bin</th>\n",
       "      <th>concentration</th>\n",
       "      <th>date</th>\n",
       "      <th>number_cells</th>\n",
       "      <th>R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin1</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200908</td>\n",
       "      <td>1440000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin1_S73_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin2</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200908</td>\n",
       "      <td>2625000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin2_S74_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin3</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200908</td>\n",
       "      <td>2370000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin3_S75_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin4</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200908</td>\n",
       "      <td>2047000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin4_S76_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib2</td>\n",
       "      <td>SortSeq_bin1</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200908</td>\n",
       "      <td>1430000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib2_RBDbin1_S77_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library        sample sample_type  sort_bin  concentration    date  \\\n",
       "0    lib1  SortSeq_bin1     SortSeq         1            NaN  200908   \n",
       "1    lib1  SortSeq_bin2     SortSeq         2            NaN  200908   \n",
       "2    lib1  SortSeq_bin3     SortSeq         3            NaN  200908   \n",
       "3    lib1  SortSeq_bin4     SortSeq         4            NaN  200908   \n",
       "4    lib2  SortSeq_bin1     SortSeq         1            NaN  200908   \n",
       "\n",
       "   number_cells  \\\n",
       "0       1440000   \n",
       "1       2625000   \n",
       "2       2370000   \n",
       "3       2047000   \n",
       "4       1430000   \n",
       "\n",
       "                                                                                                                           R1  \n",
       "0  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin1_S73_R1_001.fastq.gz  \n",
       "1  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin2_S74_R1_001.fastq.gz  \n",
       "2  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin3_S75_R1_001.fastq.gz  \n",
       "3  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin4_S76_R1_001.fastq.gz  \n",
       "4  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib2_RBDbin1_S77_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Reading Illumina runs from {config['barcode_runs']}\")\n",
    "\n",
    "illumina_runs = pd.read_csv(config['barcode_runs'])\n",
    "\n",
    "print('Here are the first few entries in the Illumina runs:')\n",
    "illumina_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for the barcode runs, initially in \"tidy\" format with one FASTQ file per row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in tidy format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin1_S73_R1_001.fastq.gz</td>\n",
       "      <td>200908_lib1_RBDbin1_S73_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin2</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin2</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin2_S74_R1_001.fastq.gz</td>\n",
       "      <td>200908_lib1_RBDbin2_S74_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin3</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin3</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin3_S75_R1_001.fastq.gz</td>\n",
       "      <td>200908_lib1_RBDbin3_S75_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin4</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin4</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin4_S76_R1_001.fastq.gz</td>\n",
       "      <td>200908_lib1_RBDbin4_S76_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib2_SortSeq_bin1</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib2_RBDbin1_S77_R1_001.fastq.gz</td>\n",
       "      <td>200908_lib2_RBDbin1_S77_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession         library_ID  \\\n",
       "0        SAMN20174027  lib1_SortSeq_bin1   \n",
       "1        SAMN20174027  lib1_SortSeq_bin2   \n",
       "2        SAMN20174027  lib1_SortSeq_bin3   \n",
       "3        SAMN20174027  lib1_SortSeq_bin4   \n",
       "4        SAMN20174027  lib2_SortSeq_bin1   \n",
       "\n",
       "                                          title library_strategy  \\\n",
       "0  Illumina barcode sequencing for SortSeq_bin1         AMPLICON   \n",
       "1  Illumina barcode sequencing for SortSeq_bin2         AMPLICON   \n",
       "2  Illumina barcode sequencing for SortSeq_bin3         AMPLICON   \n",
       "3  Illumina barcode sequencing for SortSeq_bin4         AMPLICON   \n",
       "4  Illumina barcode sequencing for SortSeq_bin1         AMPLICON   \n",
       "\n",
       "  library_source library_selection library_layout  platform  \\\n",
       "0      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "      instrument_model                 design_description filetype  \\\n",
       "0  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "1  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "2  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "3  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "4  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                                                                                            filename_fullpath  \\\n",
       "0  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin1_S73_R1_001.fastq.gz   \n",
       "1  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin2_S74_R1_001.fastq.gz   \n",
       "2  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin3_S75_R1_001.fastq.gz   \n",
       "3  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib1_RBDbin4_S76_R1_001.fastq.gz   \n",
       "4  /shared/ngs/illumina/tstarr/200917_D00300_1054_AHH332BCX3/Unaligned/Project_tstarr/200908_lib2_RBDbin1_S77_R1_001.fastq.gz   \n",
       "\n",
       "                                  filename  \n",
       "0  200908_lib1_RBDbin1_S73_R1_001.fastq.gz  \n",
       "1  200908_lib1_RBDbin2_S74_R1_001.fastq.gz  \n",
       "2  200908_lib1_RBDbin3_S75_R1_001.fastq.gz  \n",
       "3  200908_lib1_RBDbin4_S76_R1_001.fastq.gz  \n",
       "4  200908_lib2_RBDbin1_S77_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_tidy = (\n",
    "    illumina_runs\n",
    "    .assign(\n",
    "        biosample_accession=lambda x: config['biosample_accession'],\n",
    "        library_ID=lambda x: x['library'] + '_' + x['sample'],\n",
    "        title=lambda x: 'Illumina barcode sequencing for ' + x['sample'],\n",
    "        library_strategy='AMPLICON',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='PCR',\n",
    "        library_layout='single',\n",
    "        platform='ILLUMINA',\n",
    "        instrument_model='Illumina HiSeq 2500',\n",
    "        design_description='PCR of barcodes from RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['R1'].str.split(';')\n",
    "                                    .map(lambda flist: list(itertools.chain.from_iterable(glob.glob(f) for f in flist))),       \n",
    "        )\n",
    "    .explode('filename_fullpath')\n",
    "    .assign(filename_fullpath=lambda x: x['filename_fullpath'].str.strip(),\n",
    "            filename=lambda x: x['filename_fullpath'].map(os.path.basename))\n",
    "    .drop(columns=illumina_runs.columns)\n",
    "    .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "assert submissions_tidy['filename_fullpath'].map(os.path.isfile).all()\n",
    "\n",
    "print('Here are the first few submission entries in tidy format:')\n",
    "submissions_tidy.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the actual submission, we need a \"wide\" data frame that for each unique `sample_name` / `library_ID` gives all of the files each in different columns.\n",
    "These should be files without the full path.\n",
    "\n",
    "First, look at how many files there are for each sample / library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>n_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_RaACE2.787.pool6_01_bin1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib2_RaACE2.9479_05_bin4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib2_RaACE2.9479_06_bin2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib2_RaACE2.9479_06_bin3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib2_RaACE2.9479_06_bin4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_cvACE2_04_bin4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>664</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_cvACE2_05_bin1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_cvACE2_05_bin2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_cvACE2_05_bin3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib2_pgACE2_08_bin4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>668 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    biosample_accession                     library_ID  n_files\n",
       "0          SAMN20174027  lib1_RaACE2.787.pool6_01_bin1        1\n",
       "1          SAMN20174027       lib2_RaACE2.9479_05_bin4        1\n",
       "2          SAMN20174027       lib2_RaACE2.9479_06_bin2        1\n",
       "3          SAMN20174027       lib2_RaACE2.9479_06_bin3        1\n",
       "4          SAMN20174027       lib2_RaACE2.9479_06_bin4        1\n",
       "..                  ...                            ...      ...\n",
       "663        SAMN20174027            lib1_cvACE2_04_bin4        1\n",
       "664        SAMN20174027            lib1_cvACE2_05_bin1        1\n",
       "665        SAMN20174027            lib1_cvACE2_05_bin2        1\n",
       "666        SAMN20174027            lib1_cvACE2_05_bin3        1\n",
       "667        SAMN20174027            lib2_pgACE2_08_bin4        1\n",
       "\n",
       "[668 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(submissions_tidy\n",
    " .groupby(['biosample_accession', 'library_ID'])\n",
    " .aggregate(n_files=pd.NamedAgg('filename_fullpath', 'count'))\n",
    " .sort_values('n_files', ascending=False)\n",
    " .reset_index()\n",
    " )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now make the wide submission data frame.\n",
    "Note we keep only the filename column with the path lacking the full directory information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here are the first few submission entries in wide format:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_accession</th>\n",
       "      <th>library_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>200908_lib1_RBDbin1_S73_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin2</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin2</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>200908_lib1_RBDbin2_S74_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin3</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin3</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>200908_lib1_RBDbin3_S75_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib1_SortSeq_bin4</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin4</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>200908_lib1_RBDbin4_S76_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN20174027</td>\n",
       "      <td>lib2_SortSeq_bin1</td>\n",
       "      <td>Illumina barcode sequencing for SortSeq_bin1</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>200908_lib2_RBDbin1_S77_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  biosample_accession         library_ID  \\\n",
       "0        SAMN20174027  lib1_SortSeq_bin1   \n",
       "1        SAMN20174027  lib1_SortSeq_bin2   \n",
       "2        SAMN20174027  lib1_SortSeq_bin3   \n",
       "3        SAMN20174027  lib1_SortSeq_bin4   \n",
       "4        SAMN20174027  lib2_SortSeq_bin1   \n",
       "\n",
       "                                          title library_strategy  \\\n",
       "0  Illumina barcode sequencing for SortSeq_bin1         AMPLICON   \n",
       "1  Illumina barcode sequencing for SortSeq_bin2         AMPLICON   \n",
       "2  Illumina barcode sequencing for SortSeq_bin3         AMPLICON   \n",
       "3  Illumina barcode sequencing for SortSeq_bin4         AMPLICON   \n",
       "4  Illumina barcode sequencing for SortSeq_bin1         AMPLICON   \n",
       "\n",
       "  library_source library_selection library_layout  platform  \\\n",
       "0      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "      instrument_model                 design_description filetype  \\\n",
       "0  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "1  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "2  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "3  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "4  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                  filename  \n",
       "0  200908_lib1_RBDbin1_S73_R1_001.fastq.gz  \n",
       "1  200908_lib1_RBDbin2_S74_R1_001.fastq.gz  \n",
       "2  200908_lib1_RBDbin3_S75_R1_001.fastq.gz  \n",
       "3  200908_lib1_RBDbin4_S76_R1_001.fastq.gz  \n",
       "4  200908_lib2_RBDbin1_S77_R1_001.fastq.gz  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .assign(\n",
    "        filename_count=lambda x: x.groupby(['biosample_accession', 'library_ID'])['filename'].cumcount() + 1,\n",
    "        filename_col=lambda x: 'filename' + x['filename_count'].map(lambda c: str(c) if c > 1 else '')\n",
    "        )\n",
    "    .pivot(\n",
    "        index='library_ID',\n",
    "        columns='filename_col',\n",
    "        values='filename',\n",
    "        )\n",
    "    )\n",
    "\n",
    "submissions_wide = (\n",
    "    submissions_tidy\n",
    "    .drop(columns=['filename_fullpath', 'filename'])\n",
    "    .drop_duplicates()\n",
    "    .merge(submissions_wide[natsort.natsorted(submissions_wide.columns)],\n",
    "           on='library_ID',\n",
    "           validate='one_to_one',\n",
    "           )\n",
    "    )\n",
    "\n",
    "print('Here are the first few submission entries in wide format:')\n",
    "submissions_wide.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now write the wide submissions data frame to a `*.tsv` file that can be used for uploading the SRA submission website:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "submissions_spreadsheet = 'SRA_submission_spreadsheet.tsv'\n",
    "\n",
    "submissions_wide.to_csv(submissions_spreadsheet, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also want to write a file with all of the FASTQ files that are in the submission spreadsheet:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing list of all the FASTQ files to upload to FASTQs_to_upload.csv\n"
     ]
    }
   ],
   "source": [
    "fastq_file_list = 'FASTQs_to_upload.csv'\n",
    "\n",
    "print(f\"Writing list of all the FASTQ files to upload to {fastq_file_list}\")\n",
    "\n",
    "(submissions_tidy\n",
    " [['filename_fullpath', 'filename']]\n",
    " .to_csv(fastq_file_list, index=False)\n",
    " )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
